File: app/src/main/java/com/vereshchagin/nikolay/stankinschedule/ui/schedule/model/pair/Pair.java
package com.vereshchagin.nikolay.stankinschedule.ui.schedule.model.pair;

import android.os.Parcel;
import android.os.Parcelable;

import androidx.annotation.NonNull;

import com.google.gson.JsonDeserializationContext;
import com.google.gson.JsonDeserializer;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.JsonParseException;
import com.google.gson.JsonSerializationContext;
import com.google.gson.JsonSerializer;

import java.lang.reflect.Type;
import java.util.Objects;

/**
 * POJO класс пары в расписании.
 */
@Deprecated
public class Pair implements Parcelable {

    @NonNull
    private TitlePair mTitle;
    @NonNull
    private LecturerPair mLecturer;
    @NonNull
    private TypePair mType;
    @NonNull
    private SubgroupPair mSubgroup;
    @NonNull
    private ClassroomPair mClassroom;
    @NonNull
    private TimePair mTime;
    @NonNull
    private DatePair mDate;

    public Pair(@NonNull TitlePair titlePair, @NonNull LecturerPair lecturerPair,
                @NonNull TypePair typePair, @NonNull SubgroupPair subgroupPair,
                @NonNull ClassroomPair classroomPair, @NonNull TimePair timePair,
                @NonNull DatePair datePair) {

        mTitle = titlePair;
        mLecturer = lecturerPair;
        mType = typePair;
        mSubgroup = subgroupPair;
        mClassroom = classroomPair;
        mTime = timePair;
        mDate = datePair;
    }

    private Pair(@NonNull Parcel in) {
        TitlePair titlePair = in.readParcelable(TitlePair.class.getClassLoader());
        LecturerPair lecturerPair = in.readParcelable(LecturerPair.class.getClassLoader());
        TypePair typePair = in.readParcelable(TypePair.class.getClassLoader());
        SubgroupPair subgroupPair = in.readParcelable(SubgroupPair.class.getClassLoader());
        ClassroomPair classroomPair = in.readParcelable(ClassroomPair.class.getClassLoader());
        TimePair timePair = in.readParcelable(TimePair.class.getClassLoader());
        DatePair datePair = in.readParcelable(DatePair.class.getClassLoader());

        if (titlePair == null || lecturerPair == null || typePair == null
                || subgroupPair == null || classroomPair == null
                || timePair == null || datePair == null) {
            throw new IllegalArgumentException("No parsable pair: " + in);
        }

        mTitle = titlePair;
        mLecturer = lecturerPair;
        mType = typePair;
        mSubgroup = subgroupPair;
        mClassroom = classroomPair;
        mTime = timePair;
        mDate = datePair;
    }

    public static final Creator<Pair> CREATOR = new Creator<Pair>() {
        @Override
        public Pair createFromParcel(Parcel in) {
            return new Pair(in);
        }

        @Override
        public Pair[] newArray(int size) {
            return new Pair[size];
        }
    };

    @NonNull
    public TitlePair title() {
        return mTitle;
    }

    public void setTitle(@NonNull TitlePair title) {
        mTitle = title;
    }

    @NonNull
    public LecturerPair lecturer() {
        return mLecturer;
    }

    public void setLecturer(@NonNull LecturerPair lecturer) {
        mLecturer = lecturer;
    }

    @NonNull
    public TypePair type() {
        return mType;
    }

    public void setType(@NonNull TypePair type) {
        mType = type;
    }

    @NonNull
    public SubgroupPair subgroup() {
        return mSubgroup;
    }

    public void setSubgroup(@NonNull SubgroupPair subgroup) {
        mSubgroup = subgroup;
    }

    @NonNull
    public ClassroomPair classroom() {
        return mClassroom;
    }

    public void setClassroom(@NonNull ClassroomPair classroom) {
        mClassroom = classroom;
    }

    @NonNull
    public TimePair time() {
        return mTime;
    }

    public void setTime(@NonNull TimePair time) {
        mTime = time;
    }

    @NonNull
    public DatePair date() {
        return mDate;
    }

    public void setDate(@NonNull DatePair date) {
        mDate = date;
    }

    @NonNull
    @Override
    public String toString() {
        return String.format("%s. %s. %s. %s. %s. %s. %s",
                mTitle, mLecturer, mType, mSubgroup, mClassroom, mTime, mDate);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Pair pair = (Pair) o;
        return mTitle.equals(pair.mTitle) &&
                mLecturer.equals(pair.mLecturer) &&
                mType.equals(pair.mType) &&
                mSubgroup.equals(pair.mSubgroup) &&
                mClassroom.equals(pair.mClassroom) &&
                mTime.equals(pair.mTime) &&
                mDate.equals(pair.mDate);
    }

    @Override
    public int hashCode() {
        return Objects.hash(mTitle, mLecturer, mType,
                mSubgroup, mClassroom, mTime, mDate);
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeParcelable(mTitle, flags);
        dest.writeParcelable(mLecturer, flags);
        dest.writeParcelable(mType, flags);
        dest.writeParcelable(mSubgroup, flags);
        dest.writeParcelable(mClassroom, flags);
        dest.writeParcelable(mTime, flags);
        dest.writeParcelable(mDate, flags);
    }

    /**
     * Правило сериализации пары.
     */
    public static class PairSerialize implements JsonSerializer<Pair> {

        @Override
        public JsonElement serialize(Pair src, Type typeOfSrc, JsonSerializationContext context) {
            JsonObject jsonObject = new JsonObject();

            src.mTitle.toJson(jsonObject);
            src.mLecturer.toJson(jsonObject);
            src.mType.toJson(jsonObject);
            src.mSubgroup.toJson(jsonObject);
            src.mClassroom.toJson(jsonObject);
            src.mTime.toJson(jsonObject);
            src.mDate.toJson(jsonObject);

            return jsonObject;
        }
    }

    /**
     * Правило десериализиции пары.
     */
    public static class PairDeserialize implements JsonDeserializer<Pair> {

        @Override
        public Pair deserialize(JsonElement json, Type typeOfT, JsonDeserializationContext context) throws JsonParseException {
            JsonObject pairObject = json.getAsJsonObject();

            return new Pair(TitlePair.fromJson(pairObject),
                    LecturerPair.fromJson(pairObject),
                    TypePair.fromJson(pairObject),
                    SubgroupPair.fromJson(pairObject),
                    ClassroomPair.fromJson(pairObject),
                    TimePair.fromJson(pairObject),
                    DatePair.fromJson(pairObject));
        }
    }
}

File: app/src/main/java/com/vereshchagin/nikolay/stankinschedule/model/schedule/pair/Pair.kt
package com.vereshchagin.nikolay.stankinschedule.model.schedule.pair

import android.os.Parcelable
import com.google.gson.*
import kotlinx.parcelize.Parcelize

/**
 * Пара в расписании.
 * @param title название пары.
 * @param lecturer преподаватель.
 * @param classroom аудитория.
 * @param type тип пары.
 * @param subgroup подгруппа пары.
 * @param time время пары.
 * @param date даты проведения пары.
 */
@Parcelize
open class Pair(
    var title: String,
    var lecturer: String,
    var classroom: String,
    var type: Type,
    var subgroup: Subgroup,
    var time: Time,
    var date: Date,

    ) : Parcelable, Comparable<Pair> {

    constructor(
        other: Pair,
    ) : this(
        other.title,
        other.lecturer,
        other.classroom,
        other.type,
        other.subgroup,
        other.time,
        other.date
    )

    /**
     * Определяет, пересекаются ли пары по времени, дате и подгруппе.
     * @param other другая пара.
     */
    fun isIntersect(other: Pair): Boolean {
        return time.isIntersect(other.time) && date.intersect(other.date) &&
                subgroup.isIntersect(other.subgroup)
    }

    /**
     * Возвращает true, если пара может быть у этой подгруппы, иначе false.
     */
    fun isCurrently(subgroup: Subgroup): Boolean {
        return this.subgroup == Subgroup.COMMON || subgroup == Subgroup.COMMON || this.subgroup == subgroup
    }

    override fun compareTo(other: Pair): Int {
        if (time.start == other.time.start) {
            return subgroup.compareTo(other.subgroup)
        }
        return time.start.compareTo(other.time.start)
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as Pair

        if (title != other.title) return false
        if (lecturer != other.lecturer) return false
        if (classroom != other.classroom) return false
        if (type != other.type) return false
        if (subgroup != other.subgroup) return false
        if (time != other.time) return false
        if (date != other.date) return false

        return true
    }

    override fun hashCode(): Int {
        var result = title.hashCode()
        result = 31 * result + lecturer.hashCode()
        result = 31 * result + classroom.hashCode()
        result = 31 * result + type.hashCode()
        result = 31 * result + subgroup.hashCode()
        result = 31 * result + time.hashCode()
        result = 31 * result + date.hashCode()
        return result
    }

    override fun toString(): String {
        return "$title. $lecturer. $classroom. $type. $subgroup. $time. $date"
    }

    /**
     * Сериализатор пары.
     */
    class Serializer : JsonSerializer<Pair>, JsonDeserializer<Pair> {
        override fun serialize(
            src: Pair?,
            typeOfSrc: java.lang.reflect.Type?,
            context: JsonSerializationContext?,
        ): JsonElement {
            return if (src == null) {
                JsonObject()
            } else {
                JsonObject().apply {
                    addProperty(JSON_TITLE, src.title)
                    addProperty(JSON_LECTURER, src.lecturer)
                    addProperty(JSON_CLASSROOM, src.classroom)
                    addProperty(JSON_TYPE, src.type.tag)
                    addProperty(JSON_SUBGROUP, src.subgroup.tag)
                    add(JSON_TIME, src.time.toJson())
                    add(JSON_DATE, src.date.toJson())
                }
            }
        }

        override fun deserialize(
            json: JsonElement?,
            typeOfT: java.lang.reflect.Type?,
            context: JsonDeserializationContext?,
        ): Pair {
            if (json == null) {
                throw JsonParseException("Json is null")
            }

            val obj = json.asJsonObject
            return Pair(
                obj[JSON_TITLE].asString,
                obj[JSON_LECTURER].asString,
                obj[JSON_CLASSROOM].asString,
                Type.of(obj[JSON_TYPE].asString),
                Subgroup.of(obj[JSON_SUBGROUP].asString),
                Time(obj[JSON_TIME]),
                Date(obj[JSON_DATE])
            )
        }
    }

    companion object {
        const val JSON_TITLE = "title"
        const val JSON_LECTURER = "lecturer"
        const val JSON_CLASSROOM = "classroom"
        const val JSON_TYPE = "type"
        const val JSON_SUBGROUP = "subgroup"
        const val JSON_TIME = "time"
        const val JSON_DATE = "dates"
    }
}